### DISCUSSION TITLE
[Article] Mojo: The Future of AI Programming 🔥

### DISCUSSION THREAD

**Discussion by abhinav-upadhyay** [modular/modular#194] [Show and tell]
Hey everyone,

I just published an article on Mojo, focusing on Python's performance issues and how Mojo addresses them. While the Mojo team initially showcased its power with a matrix multiplication example, I opted for a simpler illustration: vector addition. Surprisingly, I discovered that an unoptimized Mojo version (using a naive for loop) was slightly faster than the equivalent Numpy code, although Numpy catches up as the vector size increases. However, it's worth noting that a SIMD optimized Mojo code outperforms Numpy by a significant margin.

I would greatly appreciate your feedback on the article.

Check it out here: [Mojo: The Future of AI Programming](https://codeconfessions.substack.com/p/mojo-the-future-of-ai-programming)

**Comment by lattner** [comment]
Very nice article!

**Reply by abhinav-upadhyay** [reply]
Thank you Chris :)

**Comment by iamtimdavis** [comment]
Thanks @abhinav-upadhyay - great article. 

One point, Mojo doesn't just come with AOT compiler, there's a JIT as well 😏 We use both and they're all important for many reasons. We also have an interpreter. 

TLDR: We have an interpreter for parametric metaprogramming system, a JIT thats part of the debugger/REPL/Workbook flow and for kernel fusion, and also AOT compiler for the Mojo CLI.

I'll close this for now and thanks for your great article!

**Reply by abhinav-upadhyay** [reply]
Thanks Tim. I didn't know about the JIT and the interpreter, I have updated the article with this information. :-) 

Also,  made another update about Python's GIL. I found out that Python is doing some work to finally have true concurrency. In Python 3.12 they are reducing the GIL scope to be at the sub-interpreter level, i.e. each sub-interpreter will have its own GIL. That means we can spawn multiple sub-interpreters in different threads and they will run concurrently. In the 3.12 release it will only be possible to do this via the C API. In 3.13 they plan to introduce a Python API.
